drivers/ethernet/phy/phy_motorcomm_yt8521.c

```diff
@@ -0,0 +1,620 @@
+{
+	const struct mc_ytphy_config *const cfg = dev->config;
+	uint16_t mask, val = 0;
+	int ret;
+
+	ret = mc_ytphy_modify_ext(dev, YT8521_CHIP_CONFIG_REG, YT8521_CCR_RXC_DLY_EN, 0);
+	if (ret) {
+		return ret;
+	}
+
+	mask = YT8521_RC1R_RX_DELAY_MASK | YT8521_RC1R_TX_DELAY_MASK;
+
+	val |= FIELD_PREP(YT8521_RC1R_RX_DELAY_MASK, cfg->rx_delay_sel);
+	val |= FIELD_PREP(YT8521_RC1R_TX_DELAY_MASK, cfg->tx_delay_sel);
+
+	return mc_ytphy_modify_ext(dev, YT8521_RGMII_CONFIG1_REG, mask, val);
+}
+
+static int mc_ytphy_resume(const struct device *dev)
+{
+	uint32_t wol_config;
+	int ret;
+
+	/* disable auto sleep */
+	ret = mc_ytphy_modify_ext(dev, YT8521_EXTREG_SLEEP_CONTROL1_REG, YT8521_ESC1R_SLEEP_SW, 0);
+	if (ret) {
+		return ret;
+	}
+
+	ret = mc_ytphy_read_ext(dev, YTPHY_WOL_CONFIG_REG, &wol_config);
+	if (ret) {
+		return ret;
+	}
+
+	/* if wol enable, do nothing */
+	if (wol_config & YTPHY_WCR_ENABLE) {
+		return 0;
+	}
+
+	return mc_ytphy_modify(dev, MII_BMCR, MII_BMCR_POWER_DOWN, 0);
+}
+
+static void invoke_link_cb(const struct device *dev)
+{
+	struct mc_ytphy_data *const data = dev->data;
+	struct phy_link_state state;
+
+	if (data->cb == NULL) {
+		return;
+	}
+
+	mc_ytphy_get_link_state(dev, &state);
+
+	data->cb(data->dev, &state, data->cb_data);
+}
+
+static inline enum phy_link_speed mc_ytphy_get_link_speed_stat_reg(const struct device *dev,
+								   uint16_t stat_reg)
+{
+	enum phy_link_speed speed;
+
+	switch (stat_reg & (SPEC_STATUS_REG_SPEED_MASK | SPEC_STATUS_REG_DUPLEX_MASK)) {
+	case PHY_SPEED_10M | PHY_DUPLEX_FULL:
+		speed = LINK_FULL_10BASE;
+		break;
+	case PHY_SPEED_10M | PHY_DUPLEX_HALF:
+		speed = LINK_HALF_10BASE;
+		break;
+	case PHY_SPEED_100M | PHY_DUPLEX_FULL:
+		speed = LINK_FULL_100BASE;
+		break;
+	case PHY_SPEED_100M | PHY_DUPLEX_HALF:
+		speed = LINK_HALF_100BASE;
+		break;
+	case PHY_SPEED_1000M | PHY_DUPLEX_FULL:
+		speed = LINK_FULL_1000BASE;
+		break;
+	case PHY_SPEED_1000M | PHY_DUPLEX_HALF:
+		speed = LINK_HALF_1000BASE;
+		break;
+	default:
+		speed = 0;
+		break;
+	}
+
+	return speed;
+}
+
+static int update_link_state(const struct device *dev)
+{
+	const struct mc_ytphy_config *const cfg = dev->config;
+	struct mc_ytphy_data *const data = dev->data;
+	uint32_t stat_reg;
+	uint32_t bmcr_reg;
+	bool link_up;
+
+	if (mc_ytphy_read(dev, YTPHY_SPECIFIC_STATUS_REG, &stat_reg) < 0) {
+		return -EIO;
+	}
+
+	link_up = (uint16_t)stat_reg & YTPHY_SSR_LINK;
+
+	/* If link is down, we can stop here. */
+	if (!link_up) {
+		data->state.speed = 0;
+		if (link_up != data->state.is_up) {
+			data->state.is_up = false;
+			LOG_INF("PHY (%d) is down", cfg->phy_addr);
+			return 0;
+		}
+		return -EAGAIN;
+	}
+
+	if (mc_ytphy_read(dev, MII_BMCR, &bmcr_reg) < 0) {
+		return -EIO;
+	}
+
+	/* If auto-negotiation is not enabled, we only need to check the link speed */
+	if ((bmcr_reg & MII_BMCR_AUTONEG_ENABLE) == 0U) {
+		enum phy_link_speed new_speed = mc_ytphy_get_link_speed_stat_reg(dev, stat_reg);
+
```

Comment:
why not using the same code of the generic phy here instead of using the `YTPHY_SPECIFIC_STATUS_REG`?

Either use the same logic as used for the generic phy or use `YTPHY_SPECIFIC_STATUS_REG` in the worker. I looked at the datasheet, if using `YTPHY_SPECIFIC_STATUS_REG` you could simplify `update_link_state` and `check_autonegotiation_completion` into one function

Reply:
Yes we could, but just like generic phy driver, it need to read difference register for the case auto negotiation is disabled and enabled,  as this phy chip provides its specific stat register, we could read this register to get speed for both cases (auto-negotiation enabled or disabled), we could call this function both in update_link_state() and check_autonegotiation_completion(), specifically it has few code in function check_autonegotiation_completion compared with generic phy driver, so I think we could use private register for this specific driver.